Source/Spray/SpraySetup.cpp

```diff
@@ -260,61 +297,115 @@ SprayParticleContainer::spraySetup(
     chemspec_names, eosparms_h);
   pele::physics::eos::speciesNames<pele::physics::PhysicsType::eos_type>(
     manivar_names, eosparms_h);
 
+  std::set<std::string> unique_chem_dep_names(
+    m_sprayDepNames, m_sprayDepNames + SPRAY_FUEL_NUM);
+  std::set<std::string> unique_manifold_dep_names(
+    m_sprayManifoldDepNames, m_sprayManifoldDepNames + SPRAY_FUEL_NUM);
+  if (unique_chem_dep_names.size() > chemspec_names.size()) {
+    amrex::Abort(
+      "Number of unique spray dep species exceeds number of chemical species "
+      "in manifold EOS");
+  }
+  if (unique_manifold_dep_names.size() != NUM_SPECIES - 1) {
+    amrex::Abort(
+      "Each liquid spray species must contribute to one manifold "
+      "parameter, as specified through dep_manifold_species");
+  }
+
   for (int i = 0; i < SPRAY_FUEL_NUM; ++i) {
-    amrex::Print() << "\n SpraySpec = " << m_sprayFuelNames[i] << " "
-                   << " dep species = " << m_sprayDepNames[i];
+    amrex::Print() << "\nSpray species " << m_sprayFuelNames[i]
+                   << " deposits to: \n";
     for (int ns = 0; ns < chemspec_names.size(); ++ns) {
-      amrex::Print() << "\n ChemSpec from manifold = " << chemspec_names[ns];
       std::string gas_spec = chemspec_names[ns];
-      if (gas_spec == m_sprayFuelNames[i]) {
-        m_sprayData->indx[i] = ns;
+      if (gas_spec == m_sprayDepNames[i]) {
+        amrex::Print() << "   Chemical species from manifold "
+                       << chemspec_names[ns] << "\n";
+        m_sprayData->dep_indx[i] = ns;
       }
     }
-    if (m_sprayData->indx[i] < 0) {
+    if (m_sprayData->dep_indx[i] < 0) {
       Abort(
-        "Fuel " + m_sprayFuelNames[i] +
+        "    Spray dep species " + m_sprayDepNames[i] +
         " not found in species available in the manifold");
     }
     for (int ns = 0; ns < manivar_names.size(); ++ns) {
-      std::string gas_spec = manivar_names[ns];
-      amrex::Print() << "\n Manivar = " << manivar_names[ns];
-      if (gas_spec == m_sprayDepNames[i]) {
-        m_sprayData->dep_indx[i] = ns;
+      std::string manifold_var = manivar_names[ns];
+      if (manifold_var == m_sprayManifoldDepNames[i]) {
+        amrex::Print() << "   Manifold var " << manivar_names[ns] << "\n";
+        m_sprayData->dep_manifold_indx[i] = ns;
       }
     }
-    if (m_sprayData->dep_indx[i] < 0) {
+    if (m_sprayData->dep_manifold_indx[i] < 0) {
       Abort(
-        "dep_fuel_species " + m_sprayDepNames[i] +
+        "    Manifold dep species " + m_sprayManifoldDepNames[i] +
         " not found as a manifold parameter");
     }
   }
 
+  // CSR representation of mapping matrix L
+  int nnz = 0; // number of non-zeros
+  m_sprayData->L_row[0] = 0;
+  for (int ns = 0; ns < chemspec_names.size(); ++ns) {
+    for (int spf = 0; spf < SPRAY_FUEL_NUM; ++spf) {
+      if (ns == m_sprayData->dep_indx[spf]) {
+        m_sprayData->L_col[nnz] = spf;
+        ++nnz;
+      }
+    }
+    m_sprayData->L_row[ns + 1] = nnz;
```

Comment:
If L_row is length NUM_SPECIES + 1, this will fail when chemspec_names (which can be arbitrary in length) is longer than NUM_SPECIES